In [1]:
import numpy as np
import pandas as pd

import os

In [2]:
DATA_PATH = '../data/'
RANDOM_SEED = 1337

In [3]:
os.listdir(DATA_PATH)

['audi.csv',
 'bmw.csv',
 'cclass.csv',
 'data.csv',
 'focus.csv',
 'ford.csv',
 'hyundi.csv',
 'merc.csv',
 'skoda.csv',
 'toyota.csv',
 'vauxhall.csv',
 'vw.csv']

In [4]:
dataframes = {}
for file in os.listdir(DATA_PATH):
    name = file.split(".")[0]
    print(name)
    dataframes[name] = pd.read_csv(os.path.join(DATA_PATH, file))
    display(dataframes[name])

audi


model  year  price transmission  mileage fuelType  tax   mpg  engineSize
0        A1  2017  12500       Manual    15735   Petrol  150  55.4         1.4
1        A6  2016  16500    Automatic    36203   Diesel   20  64.2         2.0
2        A1  2016  11000       Manual    29946   Petrol   30  55.4         1.4
3        A4  2017  16800    Automatic    25952   Diesel  145  67.3         2.0
4        A3  2019  17300       Manual     1998   Petrol  145  49.6         1.0
...     ...   ...    ...          ...      ...      ...  ...   ...         ...
10663    A3  2020  16999       Manual     4018   Petrol  145  49.6         1.0
10664    A3  2020  16999       Manual     1978   Petrol  150  49.6         1.0
10665    A3  2020  17199       Manual      609   Petrol  150  49.6         1.0
10666    Q3  2017  19499    Automatic     8646   Petrol  150  47.9         1.4
10667    Q3  2016  15999       Manual    11855   Petrol  150  47.9         1.4

[10668 rows x 9 columns]

bmw


model  year  price transmission  mileage fuelType  tax   mpg  \
0       5 Series  2014  11200    Automatic    67068   Diesel  125  57.6   
1       6 Series  2018  27000    Automatic    14827   Petrol  145  42.8   
2       5 Series  2016  16000    Automatic    62794   Diesel  160  51.4   
3       1 Series  2017  12750    Automatic    26676   Diesel  145  72.4   
4       7 Series  2014  14500    Automatic    39554   Diesel  160  50.4   
...          ...   ...    ...          ...      ...      ...  ...   ...   
10776         X3  2016  19000    Automatic    40818   Diesel  150  54.3   
10777   5 Series  2016  14600    Automatic    42947   Diesel  125  60.1   
10778   3 Series  2017  13100       Manual    25468   Petrol  200  42.8   
10779   1 Series  2014   9930    Automatic    45000   Diesel   30  64.2   
10780         X1  2017  15981    Automatic    59432   Diesel  125  57.6   

       engineSize  
0             2.0  
1             2.0  
2             3.0  
3             1.5  
4             3.0  
...           ...  
10776         2.0  
10777         2.0  
10778         2.0  
10779         2.0  
10780         2.0  

[10781 rows x 9 columns]

cclass


model  year  price transmission  mileage fuelType  engineSize
0      C Class  2020  30495    Automatic     1200   Diesel         2.0
1      C Class  2020  29989    Automatic     1000   Petrol         1.5
2      C Class  2020  37899    Automatic      500   Diesel         2.0
3      C Class  2019  30399    Automatic     5000   Diesel         2.0
4      C Class  2019  29899    Automatic     4500   Diesel         2.0
...        ...   ...    ...          ...      ...      ...         ...
3894   C Class  2017  14700       Manual    31357   Diesel         1.6
3895   C Class  2018  18500    Automatic    28248   Diesel         2.1
3896   C Class  2014  11900       Manual    48055   Diesel         2.1
3897   C Class  2014  11300    Automatic    49865   Diesel         2.1
3898   C Class  2014  14800    Automatic    55445   Diesel         2.1

[3899 rows x 7 columns]

data


Unnamed: 0    make  model  year  price transmission  mileage fuelType  \
0                0  toyota   GT86  2016  16000       Manual    24089   Petrol   
1                1  toyota   GT86  2017  15995       Manual    18615   Petrol   
2                2  toyota   GT86  2015  13998       Manual    27469   Petrol   
3                3  toyota   GT86  2017  18998       Manual    14736   Petrol   
4                4  toyota   GT86  2017  17498       Manual    36284   Petrol   
...            ...     ...    ...   ...    ...          ...      ...      ...   
108535       10663    audi     A3  2020  16999       Manual     4018   Petrol   
108536       10664    audi     A3  2020  16999       Manual     1978   Petrol   
108537       10665    audi     A3  2020  17199       Manual      609   Petrol   
108538       10666    audi     Q3  2017  19499    Automatic     8646   Petrol   
108539       10667    audi     Q3  2016  15999       Manual    11855   Petrol   

          tax   mpg  engineSize  
0       265.0  36.2         2.0  
1       145.0  36.2         2.0  
2       265.0  36.2         2.0  
3       150.0  36.2         2.0  
4       145.0  36.2         2.0  
...       ...   ...         ...  
108535  145.0  49.6         1.0  
108536  150.0  49.6         1.0  
108537  150.0  49.6         1.0  
108538  150.0  47.9         1.4  
108539  150.0  47.9         1.4  

[108540 rows x 11 columns]

focus


model  year  price transmission  mileage fuelType  engineSize
0      Focus  2016   8000       Manual    38852   Petrol         1.0
1      Focus  2019  13400       Manual    11952   Petrol         1.0
2      Focus  2019  14600       Manual    22142   Petrol         1.5
3      Focus  2016   9450       Manual    14549   Diesel         1.6
4      Focus  2015   9999       Manual     7010   Diesel         1.6
...      ...   ...    ...          ...      ...      ...         ...
5449   Focus  2019  18745       Manual     7855   Diesel         2.0
5450   Focus  2019  16350       Manual    13891   Petrol         1.0
5451   Focus  2019  16850       Manual    13452   Petrol         1.0
5452   Focus  2019  17310    Automatic    13376   Petrol         1.0
5453   Focus  2019  17610    Automatic     7886   Petrol         1.0

[5454 rows x 7 columns]

ford


model  year  price transmission  mileage fuelType  tax   mpg  \
0       Fiesta  2017  12000    Automatic    15944   Petrol  150  57.7   
1        Focus  2018  14000       Manual     9083   Petrol  150  57.7   
2        Focus  2017  13000       Manual    12456   Petrol  150  57.7   
3       Fiesta  2019  17500       Manual    10460   Petrol  145  40.3   
4       Fiesta  2019  16500    Automatic     1482   Petrol  145  48.7   
...        ...   ...    ...          ...      ...      ...  ...   ...   
17960   Fiesta  2016   7999       Manual    31348   Petrol  125  54.3   
17961    B-MAX  2017   8999       Manual    16700   Petrol  150  47.1   
17962    B-MAX  2014   7499       Manual    40700   Petrol   30  57.7   
17963    Focus  2015   9999       Manual     7010   Diesel   20  67.3   
17964       KA  2018   8299       Manual     5007   Petrol  145  57.7   

       engineSize  
0             1.0  
1             1.0  
2             1.0  
3             1.5  
4             1.0  
...           ...  
17960         1.2  
17961         1.4  
17962         1.0  
17963         1.6  
17964         1.2  

[17965 rows x 9 columns]

hyundi


model  year  price transmission  mileage fuelType  tax(£)   mpg  \
0         I20  2017   7999       Manual    17307   Petrol     145  58.9   
1      Tucson  2016  14499    Automatic    25233   Diesel     235  43.5   
2      Tucson  2016  11399       Manual    37877   Diesel      30  61.7   
3         I10  2016   6499       Manual    23789   Petrol      20  60.1   
4        IX35  2015  10199       Manual    33177   Diesel     160  51.4   
...       ...   ...    ...          ...      ...      ...     ...   ...   
4855      I30  2016   8680       Manual    25906   Diesel       0  78.4   
4856      I40  2015   7830       Manual    59508   Diesel      30  65.7   
4857      I10  2017   6830       Manual    13810   Petrol      20  60.1   
4858   Tucson  2018  13994       Manual    23313   Petrol     145  44.8   
4859   Tucson  2016  15999    Automatic    11472   Diesel     125  57.6   

      engineSize  
0            1.2  
1            2.0  
2            1.7  
3            1.0  
4            2.0  
...          ...  
4855         1.6  
4856         1.7  
4857         1.0  
4858         1.6  
4859         1.7  

[4860 rows x 9 columns]

merc


model  year  price transmission  mileage fuelType  tax   mpg  \
0             SLK  2005   5200    Automatic    63000   Petrol  325  32.1   
1         S Class  2017  34948    Automatic    27000   Hybrid   20  61.4   
2        SL CLASS  2016  49948    Automatic     6200   Petrol  555  28.0   
3         G Class  2016  61948    Automatic    16000   Petrol  325  30.4   
4         G Class  2016  73948    Automatic     4000   Petrol  325  30.1   
...           ...   ...    ...          ...      ...      ...  ...   ...   
13114     C Class  2020  35999    Automatic      500   Diesel  145  55.4   
13115     B Class  2020  24699    Automatic     2500   Diesel  145  55.4   
13116   GLC Class  2019  30999    Automatic    11612   Diesel  145  41.5   
13117   CLS Class  2019  37990    Automatic     2426   Diesel  145  45.6   
13118     S Class  2019  54999    Automatic     2075   Diesel  145  52.3   

       engineSize  
0             1.8  
1             2.1  
2             5.5  
3             4.0  
4             4.0  
...           ...  
13114         2.0  
13115         2.0  
13116         2.1  
13117         2.0  
13118         2.9  

[13119 rows x 9 columns]

skoda


model  year  price transmission  mileage fuelType  tax   mpg  \
0           Octavia  2017  10550       Manual    25250   Petrol  150  54.3   
1            Citigo  2018   8200       Manual     1264   Petrol  145  67.3   
2           Octavia  2019  15650    Automatic     6825   Diesel  145  67.3   
3      Yeti Outdoor  2015  14000    Automatic    28431   Diesel  165  51.4   
4            Superb  2019  18350       Manual    10912   Petrol  150  40.9   
...             ...   ...    ...          ...      ...      ...  ...   ...   
6262           Yeti  2014  11440    Semi-Auto    14569   Petrol  160  44.8   
6263        Octavia  2014  10990    Semi-Auto    49999   Petrol   30  56.5   
6264          Fabia  2017   9500    Semi-Auto    17131   Petrol  145  61.4   
6265         Citigo  2016   5999       Manual    21747   Petrol   20  62.8   
6266          Fabia  2017   9232    Semi-Auto    42530   Petrol   20  60.1   

      engineSize  
0            1.4  
1            1.0  
2            2.0  
3            2.0  
4            1.5  
...          ...  
6262         1.2  
6263         1.4  
6264         1.0  
6265         1.0  
6266         1.2  

[6267 rows x 9 columns]

toyota


model  year  price transmission  mileage fuelType  tax   mpg  \
0               GT86  2016  16000       Manual    24089   Petrol  265  36.2   
1               GT86  2017  15995       Manual    18615   Petrol  145  36.2   
2               GT86  2015  13998       Manual    27469   Petrol  265  36.2   
3               GT86  2017  18998       Manual    14736   Petrol  150  36.2   
4               GT86  2017  17498       Manual    36284   Petrol  145  36.2   
...              ...   ...    ...          ...      ...      ...  ...   ...   
6733              IQ  2011   5500    Automatic    30000   Petrol   20  58.9   
6734   Urban Cruiser  2011   4985       Manual    36154   Petrol  125  50.4   
6735   Urban Cruiser  2012   4995       Manual    46000   Diesel  125  57.6   
6736   Urban Cruiser  2011   3995       Manual    60700   Petrol  125  50.4   
6737   Urban Cruiser  2011   4495       Manual    45128   Petrol  125  50.4   

      engineSize  
0            2.0  
1            2.0  
2            2.0  
3            2.0  
4            2.0  
...          ...  
6733         1.0  
6734         1.3  
6735         1.4  
6736         1.3  
6737         1.3  

[6738 rows x 9 columns]

vauxhall


model  year  price transmission  mileage fuelType  tax   mpg  \
0        Corsa  2018   7885       Manual     9876   Petrol  145  55.4   
1        Corsa  2019  11995       Manual     2500   Petrol  145  54.3   
2        Corsa  2017   9777    Automatic     9625   Petrol  145  47.9   
3        Corsa  2016   8500       Manual    25796   Petrol   30  55.4   
4        Corsa  2019  10000       Manual     3887   Petrol  145  43.5   
...        ...   ...    ...          ...      ...      ...  ...   ...   
13627    Tigra  2007   2499       Manual    48640   Petrol  165  46.3   
13628   Vectra  2007   1795       Manual    75000   Petrol  240  39.2   
13629   Vectra  2005   1495       Manual    74000   Petrol  235  38.7   
13630   Vectra  2008   1695    Automatic   131000   Diesel  200  39.8   
13631   Vectra  2005   1450       Manual   147000   Petrol  330  27.9   

       engineSize  
0             1.4  
1             1.4  
2             1.4  
3             1.4  
4             1.4  
...           ...  
13627         1.4  
13628         1.8  
13629         1.8  
13630         1.9  
13631         3.2  

[13632 rows x 9 columns]

vw


model  year  price transmission  mileage fuelType  tax   mpg  \
0       T-Roc  2019  25000    Automatic    13904   Diesel  145  49.6   
1       T-Roc  2019  26883    Automatic     4562   Diesel  145  49.6   
2       T-Roc  2019  20000       Manual     7414   Diesel  145  50.4   
3       T-Roc  2019  33492    Automatic     4825   Petrol  145  32.5   
4       T-Roc  2019  22900    Semi-Auto     6500   Petrol  150  39.8   
...       ...   ...    ...          ...      ...      ...  ...   ...   
15152     Eos  2012   5990       Manual    74000   Diesel  125  58.9   
15153     Fox  2008   1799       Manual    88102   Petrol  145  46.3   
15154     Fox  2009   1590       Manual    70000   Petrol  200  42.0   
15155     Fox  2006   1250       Manual    82704   Petrol  150  46.3   
15156     Fox  2007   2295       Manual    74000   Petrol  145  46.3   

       engineSize  
0             2.0  
1             2.0  
2             2.0  
3             2.0  
4             1.5  
...           ...  
15152         2.0  
15153         1.2  
15154         1.4  
15155         1.2  
15156         1.2  

[15157 rows x 9 columns]

In [5]:
dataframes["merc"] = dataframes["merc"].append(dataframes["cclass"])
dataframes["ford"] = dataframes["ford"].append(dataframes["focus"])
del dataframes["focus"], dataframes["cclass"]
dataframes["ford"]


model  year  price transmission  mileage fuelType    tax   mpg  \
0      Fiesta  2017  12000    Automatic    15944   Petrol  150.0  57.7   
1       Focus  2018  14000       Manual     9083   Petrol  150.0  57.7   
2       Focus  2017  13000       Manual    12456   Petrol  150.0  57.7   
3      Fiesta  2019  17500       Manual    10460   Petrol  145.0  40.3   
4      Fiesta  2019  16500    Automatic     1482   Petrol  145.0  48.7   
...       ...   ...    ...          ...      ...      ...    ...   ...   
5449    Focus  2019  18745       Manual     7855   Diesel    NaN   NaN   
5450    Focus  2019  16350       Manual    13891   Petrol    NaN   NaN   
5451    Focus  2019  16850       Manual    13452   Petrol    NaN   NaN   
5452    Focus  2019  17310    Automatic    13376   Petrol    NaN   NaN   
5453    Focus  2019  17610    Automatic     7886   Petrol    NaN   NaN   

      engineSize  
0            1.0  
1            1.0  
2            1.0  
3            1.5  
4            1.0  
...          ...  
5449         2.0  
5450         1.0  
5451         1.0  
5452         1.0  
5453         1.0  

[23419 rows x 9 columns]

In [15]:
dataframes["merc"]

model  year  price transmission  mileage fuelType    tax   mpg  \
0           SLK  2005   5200    Automatic    63000   Petrol  325.0  32.1   
1       S Class  2017  34948    Automatic    27000   Hybrid   20.0  61.4   
2      SL CLASS  2016  49948    Automatic     6200   Petrol  555.0  28.0   
3       G Class  2016  61948    Automatic    16000   Petrol  325.0  30.4   
4       G Class  2016  73948    Automatic     4000   Petrol  325.0  30.1   
...         ...   ...    ...          ...      ...      ...    ...   ...   
3894    C Class  2017  14700       Manual    31357   Diesel    NaN   NaN   
3895    C Class  2018  18500    Automatic    28248   Diesel    NaN   NaN   
3896    C Class  2014  11900       Manual    48055   Diesel    NaN   NaN   
3897    C Class  2014  11300    Automatic    49865   Diesel    NaN   NaN   
3898    C Class  2014  14800    Automatic    55445   Diesel    NaN   NaN   

      engineSize  
0            1.8  
1            2.1  
2            5.5  
3            4.0  
4            4.0  
...          ...  
3894         1.6  
3895         2.1  
3896         2.1  
3897         2.1  
3898         2.1  

[17018 rows x 9 columns]

In [6]:
df_merc_duplicated = dataframes["merc"][dataframes["merc"].duplicated(subset=['model', 'year', 'price', 'transmission', 'mileage', 'fuelType', 'engineSize'])]

df_merc_duplicated

model  year  price transmission  mileage fuelType    tax   mpg  \
493      V Class  2018  19998       Manual    27222   Diesel  145.0  46.3   
494      V Class  2016  17998       Manual    21981   Diesel  200.0  46.3   
633      C Class  2020  29950    Semi-Auto      101   Diesel  145.0  48.7   
1328   GLB Class  2020  35400    Semi-Auto     2000   Diesel  145.0  46.3   
1505     E Class  2016  15995    Semi-Auto    60931   Diesel  125.0  58.9   
...          ...   ...    ...          ...      ...      ...    ...   ...   
3894     C Class  2017  14700       Manual    31357   Diesel    NaN   NaN   
3895     C Class  2018  18500    Automatic    28248   Diesel    NaN   NaN   
3896     C Class  2014  11900       Manual    48055   Diesel    NaN   NaN   
3897     C Class  2014  11300    Automatic    49865   Diesel    NaN   NaN   
3898     C Class  2014  14800    Automatic    55445   Diesel    NaN   NaN   

      engineSize  
493          2.1  
494          2.1  
633          2.0  
1328         2.0  
1505         2.1  
...          ...  
3894         1.6  
3895         2.1  
3896         2.1  
3897         2.1  
3898         2.1  

[3936 rows x 9 columns]

Hay 3936 vehículos duplicados en el dataset de mercedes.

In [7]:
dataframes["merc"].drop_duplicates(subset=['model', 'year', 'price', 'transmission', 'mileage', 'fuelType', 'engineSize'],inplace=True)
dataframes["merc"].count()

model           13082
year            13082
price           13082
transmission    13082
mileage         13082
fuelType        13082
tax             12818
mpg             12818
engineSize      13082
dtype: int64

In [8]:
df_merc_post_duplicated = dataframes["merc"][dataframes["merc"].duplicated(subset=['model', 'year', 'price', 'transmission', 'mileage', 'fuelType', 'engineSize'])]

df_merc_post_duplicated

Empty DataFrame
Columns: [model, year, price, transmission, mileage, fuelType, tax, mpg, engineSize]
Index: []

In [9]:
df_ford_duplicated = dataframes["ford"][dataframes["ford"].duplicated(subset=['model', 'year', 'price', 'transmission', 'mileage', 'fuelType', 'engineSize'])]

df_ford_duplicated

model  year  price transmission  mileage fuelType    tax   mpg  \
159       Focus  2016  10995       Manual    30923   Diesel    0.0  74.3   
311       Focus  2019  14998       Manual     7000   Petrol  145.0  60.1   
312       Focus  2019  14998       Manual     7000   Petrol  145.0  58.9   
349       Focus  2019  14998       Manual     7000   Petrol  145.0  58.9   
510    EcoSport  2019  15489       Manual     5636   Diesel  145.0  68.9   
...         ...   ...    ...          ...      ...      ...    ...   ...   
5449      Focus  2019  18745       Manual     7855   Diesel    NaN   NaN   
5450      Focus  2019  16350       Manual    13891   Petrol    NaN   NaN   
5451      Focus  2019  16850       Manual    13452   Petrol    NaN   NaN   
5452      Focus  2019  17310    Automatic    13376   Petrol    NaN   NaN   
5453      Focus  2019  17610    Automatic     7886   Petrol    NaN   NaN   

      engineSize  
159          1.5  
311          1.0  
312          1.0  
349          1.0  
510          1.5  
...          ...  
5449         2.0  
5450         1.0  
5451         1.0  
5452         1.0  
5453         1.0  

[5202 rows x 9 columns]

In [10]:
dataframes["ford"].drop_duplicates(subset=['model', 'year', 'price', 'transmission', 'mileage', 'fuelType', 'engineSize'], inplace=True)

dataframes["ford"].count()

model           18217
year            18217
price           18217
transmission    18217
mileage         18217
fuelType        18217
tax             17783
mpg             17783
engineSize      18217
dtype: int64

**Hasta aquí el dataframe de ford y mercedes no contiene datos duplicados, veamos el resto de dataframes de los otros modelos**

In [26]:
for make in dataframes:

    if make not in ["merc", "ford"]:
        duplicates = dataframes[make].duplicated().sum()

        if duplicates > 0:
            dataframes[make].drop_duplicates(inplace=True)
    if(dataframes[make].duplicated().sum()) != 0:
        display(make) #check

In [5]:
dataframes["hyundi"].rename(columns={"tax(£)": "tax"}, inplace=True)
dataframes["hyundi"]

model  year  price transmission  mileage fuelType  tax   mpg  \
0         I20  2017   7999       Manual    17307   Petrol  145  58.9   
1      Tucson  2016  14499    Automatic    25233   Diesel  235  43.5   
2      Tucson  2016  11399       Manual    37877   Diesel   30  61.7   
3         I10  2016   6499       Manual    23789   Petrol   20  60.1   
4        IX35  2015  10199       Manual    33177   Diesel  160  51.4   
...       ...   ...    ...          ...      ...      ...  ...   ...   
4855      I30  2016   8680       Manual    25906   Diesel    0  78.4   
4856      I40  2015   7830       Manual    59508   Diesel   30  65.7   
4857      I10  2017   6830       Manual    13810   Petrol   20  60.1   
4858   Tucson  2018  13994       Manual    23313   Petrol  145  44.8   
4859   Tucson  2016  15999    Automatic    11472   Diesel  125  57.6   

      engineSize  
0            1.2  
1            2.0  
2            1.7  
3            1.0  
4            2.0  
...          ...  
4855         1.6  
4856         1.7  
4857         1.0  
4858         1.6  
4859         1.7  

[4860 rows x 9 columns]

In [6]:
columns = [
    "make", "model", "year", "price", "transmission", "mileage", "fuelType",
    "tax", "mpg", "engineSize"
]
df = pd.DataFrame(columns=columns)

for name in dataframes:

    if name not in ['cclass', 'focus']:
        file_df = dataframes[name]
        file_df["make"] = name
            
        df = df.append(file_df)
df

make model  year  price transmission mileage fuelType  tax   mpg  \
0      audi    A1  2017  12500       Manual   15735   Petrol  150  55.4   
1      audi    A6  2016  16500    Automatic   36203   Diesel   20  64.2   
2      audi    A1  2016  11000       Manual   29946   Petrol   30  55.4   
3      audi    A4  2017  16800    Automatic   25952   Diesel  145  67.3   
4      audi    A3  2019  17300       Manual    1998   Petrol  145  49.6   
...     ...   ...   ...    ...          ...     ...      ...  ...   ...   
15152    vw   Eos  2012   5990       Manual   74000   Diesel  125  58.9   
15153    vw   Fox  2008   1799       Manual   88102   Petrol  145  46.3   
15154    vw   Fox  2009   1590       Manual   70000   Petrol  200  42.0   
15155    vw   Fox  2006   1250       Manual   82704   Petrol  150  46.3   
15156    vw   Fox  2007   2295       Manual   74000   Petrol  145  46.3   

       engineSize  Unnamed: 0  
0             1.4         NaN  
1             2.0         NaN  
2             1.4         NaN  
3             2.0         NaN  
4             1.0         NaN  
...           ...         ...  
15152         2.0         NaN  
15153         1.2         NaN  
15154         1.4         NaN  
15155         1.2         NaN  
15156         1.2         NaN  

[207727 rows x 11 columns]

In [19]:
df[df.isnull().any(axis=1)]["make"].unique()

array(['ford', 'merc'], dtype=object)

In [20]:
df.isnull().sum()

make               0
model              0
year               0
price              0
transmission       0
mileage            0
fuelType           0
tax             9353
mpg             9353
engineSize         0
dtype: int64

In [21]:
df[df["make"] == "ford"]["model"].value_counts()

 Focus                    10042
 Fiesta                    6557
 Kuga                      2225
 EcoSport                  1143
 C-MAX                      543
 Ka+                        531
 Mondeo                     526
 B-MAX                      355
 S-MAX                      296
 Grand C-MAX                247
 Galaxy                     228
 Edge                       208
 KA                         199
 Puma                        80
 Tourneo Custom              69
 Grand Tourneo Connect       59
 Mustang                     57
 Tourneo Connect             33
 Fusion                      16
 Streetka                     2
 Escort                       1
 Ranger                       1
 Transit Tourneo              1
Name: model, dtype: int64

In [22]:
df.groupby(["make", "model", "year"]).engineSize.unique()

make  model  year
audi   A1    2010                   [1.4]
             2011              [1.6, 1.4]
             2012    [1.6, 1.4, 2.0, 1.2]
             2013    [1.6, 1.4, 2.0, 1.2]
             2014    [1.4, 1.6, 1.2, 2.0]
                             ...         
vw     Up    2016                   [1.0]
             2017              [1.0, 0.0]
             2018                   [1.0]
             2019                   [1.0]
             2020                   [1.0]
Name: engineSize, Length: 1615, dtype: object

In [20]:
df.duplicated().sum()

NameError: name 'df' is not defined

In [38]:
df.isnull().sum()

make               0
model              0
year               0
price              0
transmission       0
mileage            0
fuelType           0
tax             9353
mpg             9353
engineSize         0
dtype: int64

In [1]:
a = ("a", 2, 3, 4)

In [2]:
a

('a', 2, 3, 4)

In [ ]:
*a --> "a", 2, 3, 4

In [3]:
b = {"key1": "value1", "key2": "value2"}
b

{'key1': 'value1', 'key2': 'value2'}

In [ ]:
**b --> key1="value1", key2="value2"

In [4]:
c = {"key3": "value3"}

In [5]:
{**b, **c}

{'key1': 'value1', 'key2': 'value2', 'key3': 'value3'}

In [7]:
dict(key1="value1", key2="value2", key3="value3")

{'key1': 'value1', 'key2': 'value2', 'key3': 'value3'}